### PR TITLE
Enable SolverIDR for block vectors

### PR DESCRIPTION
**Pull request by peterrum**

### CODE REVIEW

File: include/deal.II/lac/solver_idr.h [hunk 1]
```diff
@@ -226,6 +226,54 @@ namespace internal
...
+
+    template <typename VectorType,
+              std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
+                nullptr>
+    typename VectorType::BlockType &
+    block(VectorType &vector, const unsigned int b)
+    {
+      AssertThrow(b == 1, ExcInternalError());
+      return vector.block(b);
+    }
```
Comment (kronbichler) [inline include/deal.II/lac/solver_idr.h:275]: This code is the same as https://github.com/dealii/dealii/blob/2b2da6719b3cdea2fceea7e21a0cb510fd2b1fe3/include/deal.II/matrix_free/operators.h#L44-L91 - shouldn't we try to unify this e.g. in the file https://github.com/dealii/dealii/blob/master/include/deal.II/lac/block_vector_base.h ?

File: include/deal.II/lac/solver_idr.h [hunk 2]
```diff
@@ -226,6 +226,54 @@ namespace internal
...
+
+
+
+    template <typename VectorType,
+              std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
+                nullptr>
+    typename VectorType::BlockType &
+    block(VectorType &vector, const unsigned int b)
+    {
+      AssertThrow(b == 1, ExcInternalError());
```
Comment (kronbichler) [inline include/deal.II/lac/solver_idr.h:273]: By the way, this does not look right.

**Comment by peterrum** [comment]
> This code is the same as https://github.com/dealii/dealii/blob/2b2da6719b3cdea2fceea7e21a0cb510fd2b1fe3/include/deal.II/matrix_free/operators.h#L44-L91 - shouldn't we try to unify this e.g. in the file https://github.com/dealii/dealii/blob/master/include/deal.II/lac/block_vector_base.h ?

You are right but I would suggest to keep separately and create an issue. I am just not sure if having free functions is the right approach. In the context of `Triangulation`/`DoFHandler` we have the function `get_triangulation()` with the comment

> Return a reference to the current object.
This doesn't seem to be very useful but allows to write code that can access the underlying triangulation for anything that satisfies the [MeshType concept](https://www.dealii.org/developer/doxygen/deal.II/group__Concepts.html#ConceptMeshType) (which may not only be a triangulation, but also a [DoFHandler](https://www.dealii.org/developer/doxygen/deal.II/classDoFHandler.html), for example).

(see https://www.dealii.org/developer/doxygen/deal.II/classTriangulation.html#acbcac8668993d582f192d3c85b642dd3).

Why not do the same thing here? I don't have a preference but the solution should be at least consistent.

**Comment by kronbichler** [comment]
Yes, we can open an issue.

> Why not do the same thing here? I don't have a preference but the solution should be at least consistent.

The problem is that not all vectors have a common base class, so we would need to do it in several places, and we would need to make sure that we do not shadow one or the other function. Given this, it depends on the frequency and importance of use in my opinion. Having in the vector classes might be useful if there is broad consensus. Otherwise, I don't see a problem with having free (internal) functions for less important things along this direction.

**Review by kronbichler — APPROVED** [review]

**Comment by peterrum** [comment]
@kronbichler I have opened a new issue: https://github.com/dealii/dealii/issues/14233

**Comment by masterleinad** [comment]
```
In file included from /jenkins/workspace/dealii_PR-14232/tests/all-headers/test_header.cc:17:0:
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:233:33: error: ‘IsBlockVector’ was not declared in this scope
               std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
                                 ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:233:57: error: template argument 1 is invalid
               std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
                                                         ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:234:19: error: expected unqualified-id before ‘=’ token
                 * = nullptr>
                   ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:234:19: error: expected constructor, destructor, or type conversion before ‘=’ token
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:244:32: error: ‘IsBlockVector’ was not declared in this scope
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:244:56: error: template argument 1 is invalid
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                                        ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:244:80: error: expected unqualified-id before ‘=’ token
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                                                                ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:244:80: error: expected constructor, destructor, or type conversion before ‘=’ token
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:255:33: error: ‘IsBlockVector’ was not declared in this scope
               std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
                                 ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:255:57: error: template argument 1 is invalid
               std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
                                                         ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:256:19: error: expected unqualified-id before ‘=’ token
                 * = nullptr>
                   ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:256:19: error: expected constructor, destructor, or type conversion before ‘=’ token
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:268:32: error: ‘IsBlockVector’ was not declared in this scope
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:268:56: error: template argument 1 is invalid
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                                        ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:268:80: error: expected unqualified-id before ‘=’ token
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                                                                ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:268:80: error: expected constructor, destructor, or type conversion before ‘=’ token
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h: In member function ‘void dealii::SolverIDR<VectorType>::solve(const MatrixType&, VectorType&, const VectorType&, const PreconditionerType&)’:
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:379:20: error: ‘n_blocks’ is not a member of ‘dealii::internal::SolverIDRImplementation’
                b < internal::SolverIDRImplementation::n_blocks(tmp_q);
                    ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:381:30: error: ‘block’ is not a member of ‘dealii::internal::SolverIDRImplementation’
             for (auto indx : internal::SolverIDRImplementation::block(tmp_q, b)
                              ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:383:15: error: ‘block’ is not a member of ‘dealii::internal::SolverIDRImplementation’
               internal::SolverIDRImplementation::block(tmp_q, b)(indx) =
               ^
make[3]: *** [CMakeFiles/lac-solver_idr.h.debug.dir/test_header.cc.o] Error 1
make[2]: *** [CMakeFiles/lac-solver_idr.h.debug.dir/all] Error 2
make[1]: *** [CMakeFiles/lac-solver_idr.h.debug.build.dir/rule] Error 2
make: *** [lac-solver_idr.h.debug.build] Error 2
Test all-headers/lac/solver_idr.h.debug: BUILD
===============================   OUTPUT BEGIN  ===============================
all-headers/lac/solver_idr.h.debug: BUILD failed. Output:
Generating lac-solver_idr.h.debug_interrupt_guard.cc
Scanning dependencies of target lac-solver_idr.h.debug
Building CXX object CMakeFiles/lac-solver_idr.h.debug.dir/test_header.cc.o
CMakeFiles/lac-solver_idr.h.debug.dir/build.make:66: recipe for target 'CMakeFiles/lac-solver_idr.h.debug.dir/test_header.cc.o' failed
CMakeFiles/Makefile2:20803: recipe for target 'CMakeFiles/lac-solver_idr.h.debug.dir/all' failed
CMakeFiles/Makefile2:15530: recipe for target 'CMakeFiles/lac-solver_idr.h.debug.build.dir/rule' failed
Makefile:6408: recipe for target 'lac-solver_idr.h.debug.build' failed
all-headers/lac/solver_idr.h.debug: ******    BUILD failed    *******
===============================    OUTPUT END   ===============================
Expected stage PASSED - aborting
CMake Error at /home/dealii/build/share/deal.II/scripts/run_test.cmake:144 (MESSAGE):
  *** abort
Standard Output
In file included from /jenkins/workspace/dealii_PR-14232/tests/all-headers/test_header.cc:17:0:
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:233:33: error: ‘IsBlockVector’ was not declared in this scope
               std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
                                 ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:233:57: error: template argument 1 is invalid
               std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
                                                         ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:234:19: error: expected unqualified-id before ‘=’ token
                 * = nullptr>
                   ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:234:19: error: expected constructor, destructor, or type conversion before ‘=’ token
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:244:32: error: ‘IsBlockVector’ was not declared in this scope
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:244:56: error: template argument 1 is invalid
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                                        ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:244:80: error: expected unqualified-id before ‘=’ token
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                                                                ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:244:80: error: expected constructor, destructor, or type conversion before ‘=’ token
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:255:33: error: ‘IsBlockVector’ was not declared in this scope
               std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
                                 ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:255:57: error: template argument 1 is invalid
               std::enable_if_t<!IsBlockVector<VectorType>::value, VectorType>
                                                         ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:256:19: error: expected unqualified-id before ‘=’ token
                 * = nullptr>
                   ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:256:19: error: expected constructor, destructor, or type conversion before ‘=’ token
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:268:32: error: ‘IsBlockVector’ was not declared in this scope
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:268:56: error: template argument 1 is invalid
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                                        ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:268:80: error: expected unqualified-id before ‘=’ token
               std::enable_if_t<IsBlockVector<VectorType>::value, VectorType> * =
                                                                                ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:268:80: error: expected constructor, destructor, or type conversion before ‘=’ token
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h: In member function ‘void dealii::SolverIDR<VectorType>::solve(const MatrixType&, VectorType&, const VectorType&, const PreconditionerType&)’:
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:379:20: error: ‘n_blocks’ is not a member of ‘dealii::internal::SolverIDRImplementation’
                b < internal::SolverIDRImplementation::n_blocks(tmp_q);
                    ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:381:30: error: ‘block’ is not a member of ‘dealii::internal::SolverIDRImplementation’
             for (auto indx : internal::SolverIDRImplementation::block(tmp_q, b)
                              ^
/jenkins/workspace/dealii_PR-14232/include/deal.II/lac/solver_idr.h:383:15: error: ‘block’ is not a member of ‘dealii::internal::SolverIDRImplementation’
               internal::SolverIDRImplementation::block(tmp_q, b)(indx) =
               ^
make[3]: *** [CMakeFiles/lac-solver_idr.h.debug.dir/test_header.cc.o] Error 1
make[2]: *** [CMakeFiles/lac-solver_idr.h.debug.dir/all] Error 2
make[1]: *** [CMakeFiles/lac-solver_idr.h.debug.build.dir/rule] Error 2
make: *** [lac-solver_idr.h.debug.build] Error 2
Test all-headers/lac/solver_idr.h.debug: BUILD
===============================   OUTPUT BEGIN  ===============================
all-headers/lac/solver_idr.h.debug: BUILD failed. Output:
Generating lac-solver_idr.h.debug_interrupt_guard.cc
Scanning dependencies of target lac-solver_idr.h.debug
Building CXX object CMakeFiles/lac-solver_idr.h.debug.dir/test_header.cc.o
CMakeFiles/lac-solver_idr.h.debug.dir/build.make:66: recipe for target 'CMakeFiles/lac-solver_idr.h.debug.dir/test_header.cc.o' failed
CMakeFiles/Makefile2:20803: recipe for target 'CMakeFiles/lac-solver_idr.h.debug.dir/all' failed
CMakeFiles/Makefile2:15530: recipe for target 'CMakeFiles/lac-solver_idr.h.debug.build.dir/rule' failed
Makefile:6408: recipe for target 'lac-solver_idr.h.debug.build' failed
all-headers/lac/solver_idr.h.debug: ******    BUILD failed    *******
===============================    OUTPUT END   ===============================
Expected stage PASSED - aborting
CMake Error at /home/dealii/build/share/deal.II/scripts/run_test.cmake:144 (MESSAGE):
  *** abort
```

**Review by kronbichler — APPROVED** [review]

**Pull request merged**